File: ios/Classes/FeaturesStreamHandler.swift
//
//  FeaturesStreamHandler.swift
//  mapbox_gl
//
//  Created by Bjarte Bore on 15/11/2021.
//

import Foundation
import MapboxMaps


class FeaturesStreamHandler : NSObject, FlutterStreamHandler {
    var client: MapView
    var events: FlutterEventSink?
    var source: String
    var sourceLayers: [String]?
    var filter: Any
    var layerId: String?
    
    init(client: MapView, source: String, sourceLayers: [String]?, filter: Any) {
        self.client = client
        self.source = source
        self.sourceLayers = sourceLayers
        self.filter = filter
    }

    func dispatchFeatures() {
        let queryOptions = SourceQueryOptions(sourceLayerIds: sourceLayers, filter: filter)
        
        client.mapboxMap.querySourceFeatures(for: source, options: queryOptions) {
            result in
            
            if case .success(let features) = result,
               features.count > 0 {
                let rawFeatures = features.map { feature in
                    return feature.feature;
                }
                if let data = try? JSONEncoder().encode(rawFeatures) {
                    let result = String(data: data, encoding: String.Encoding.utf8);
                    self.events?(result)
                }
            }
       }
    }
        
    
    func onMapDataLoaded(event: Event) {
        if let data = event.data as? [String: Any] {
            if let id = data["id"] as? String {
                if(id == self.source) {
                    dispatchFeatures()
                }
            }
        }
    }
    
    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        
        let args = arguments as! Dictionary<String, Any>

        layerId = args["layerId"] as? String

        self.events = events;
        
        client.mapboxMap.onEvery(.sourceDataLoaded, handler: onMapDataLoaded)

        dispatchFeatures()
        
        return nil
    }
    
    
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.events = nil
        
        return nil;
    }
}

File: android/src/main/kotlin/com/mapbox/mapboxgl/FeaturesStreamHandler.kt
package com.mapbox.mapboxgl

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.mapbox.bindgen.Value
import com.mapbox.maps.MapView
import com.mapbox.maps.QueriedFeature
import com.mapbox.maps.QueryFeaturesCallback
import com.mapbox.maps.SourceQueryOptions
import com.mapbox.maps.extension.observable.eventdata.SourceDataLoadedEventData
import com.mapbox.maps.plugin.delegates.listeners.OnSourceDataLoadedListener
import io.flutter.plugin.common.EventChannel
import java.util.HashMap

class FeaturesStreamHandler(
        client: MapView,
        source: String,
        sourceLayers: List<String>?,
        filter: Any?
) : EventChannel.StreamHandler, OnSourceDataLoadedListener {
    private var events: EventChannel.EventSink? = null
    private val client: MapView = client
    private val source: String = source
    private val sourceLayers: List<String>? = sourceLayers
    private val filter: Any? = filter
    private var layerId: String? = null
    private val gson = Gson()

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        (arguments as Map<String, Any?>?)?.let { args ->
            layerId = args["layerId"] as String?
            this.events = events;

            client.getMapboxMap().addOnSourceDataLoadedListener(this)

            dispatchFeatures();
        }
    }

    override fun onCancel(arguments: Any?) {
        this.events = null
    }

    private fun getFilter(): Value {
        if (filter is Double) {
            return Value.valueOf(filter)
        }
        if (filter is Long) {
            return Value.valueOf(filter)
        }
        if (filter is Boolean) {
            return Value.valueOf(filter)
        }
        if (filter is List<*> || filter is HashMap<*, *> ) {
            val json = gson.toJson(filter);
            val expected = Value.fromJson(json)
            if (expected.isValue) {
                return expected.value!!
            }
        }

        return Value.nullValue()
    }

    private fun dispatchFeatures() {
        val queryOptions = SourceQueryOptions(sourceLayers, getFilter())

        this.events?.let { events ->

            client.getMapboxMap().querySourceFeatures(source, queryOptions, QueryFeaturesCallback { result ->

                ( result.value as List<QueriedFeature>?)?.let { qFeatures ->
                    val rawFeaturess = qFeatures.map<QueriedFeature, Map<String, Any>> {  feature ->
                        val json = gson.toJson(feature.feature);

                        return@map gson.fromJson(
                                json,
                                object : TypeToken<HashMap<String?, Any?>?>() {}.type
                        );
                    }
                    events.success(rawFeaturess )

                }
            })

        }

    }
    override fun onSourceDataLoaded(eventData: SourceDataLoadedEventData) {
        if (eventData.id == source) {
            dispatchFeatures()
        }
    }

}